fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/LayeredDrawerWrapperImpl.java

```diff
@@ -0,0 +1,179 @@
+
+import org.jetbrains.annotations.VisibleForTesting;
+
+import net.minecraft.client.gui.LayeredDrawer;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.client.rendering.v1.IdentifiedLayer;
+import net.fabricmc.fabric.api.client.rendering.v1.LayeredDrawerWrapper;
+import net.fabricmc.fabric.mixin.client.rendering.LayeredDrawerAccessor;
+
+public final class LayeredDrawerWrapperImpl implements LayeredDrawerWrapper {
+	private final LayeredDrawer base;
+
+	public LayeredDrawerWrapperImpl(LayeredDrawer base) {
+		this.base = base;
+	}
+
+	private static List<LayeredDrawer.Layer> getLayers(LayeredDrawer drawer) {
+		return ((LayeredDrawerAccessor) drawer).getLayers();
+	}
+
+	@Override
+	public LayeredDrawerWrapper addLayer(IdentifiedLayer layer) {
+		validateUnique(layer);
+		getLayers(this.base).add(layer);
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper attachLayerAfter(Identifier afterThis, IdentifiedLayer layer) {
+		validateUnique(layer);
+
+		boolean didChange = findLayer(afterThis, (l, iterator) -> {
+			iterator.add(layer);
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + afterThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper attachLayerBefore(Identifier beforeThis, IdentifiedLayer layer) {
+		validateUnique(layer);
+		boolean didChange = findLayer(beforeThis, (l, iterator) -> {
+			iterator.previous();
+			iterator.add(layer);
+			iterator.next();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + beforeThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper removeLayer(Identifier identifier) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper replaceLayer(Identifier identifier, Function<IdentifiedLayer, IdentifiedLayer> replacer) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			iterator.add(replacer.apply((IdentifiedLayer) l));
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	@VisibleForTesting
+	void validateUnique(IdentifiedLayer layer) {
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, layer.id())) {
+				throw new IllegalArgumentException("Layer with identifier " + layer.id() + " already exists");
+			}
+
+			return false;
+		});
+	}
+
+	/**
+	 * @return true if a layer with the given identifier was found
+	 */
+	@VisibleForTesting
+	boolean findLayer(Identifier identifier, LayerVisitor visitor) {
+		AtomicBoolean found = new AtomicBoolean(false);
```

Comment:
Imo it's better to use apache's `MutableBoolean` than `AtomicBoolean` for mutating from within a lambda, unless the lambda is called from multiple threads.

Reply:
I agree; this change should be made.

Reply:
Also on line 149.

Reply:
Done in e3f65ea.